docs/source/getting_started/benchmarks/cifar10_benchmark.py

```diff
@@ -729,17 +729,113 @@ def configure_optimizers(self):
         return [optim], [scheduler]
 
 
+# import here so as to not have an additional dependency
+# TODO: replace by faiss?
+from sklearn.cluster import KMeans
+
+class SMoGModel(BenchmarkModule):
+
+    def __init__(self, dataloader_kNN, num_classes):
+        super().__init__(dataloader_kNN, num_classes)
+
+        # create a ResNet backbone and remove the classification head
+        resnet = lightly.models.ResNetGenerator('resnet-18')
+        self.backbone = nn.Sequential(
+            *list(resnet.children())[:-1],
+            nn.AdaptiveAvgPool2d(1)
+        )
+
+        # create a moco model based on ResNet
+        self.projection_head = heads.SMoGProjectionHead(512, 2048, 128)
+        self.prediction_head = heads.ProjectionHead([
+            (128, 2048, nn.BatchNorm1d(2048), nn.ReLU()),
+            (2048, 128, None, nn.ReLU())
+        ])
+        self._reset_momentum_weights()
+
+        # smog
+        self.n_groups = 3000 # 2% malus vs optimal setting of 3000 groups
+        self.memory_bank = lightly.loss.memory_bank.MemoryBankModule(size=10000)
+        # create our loss
+        self.smog = modules.SMoG(self.n_groups, 128, 0.99, device='cuda')
+        self.criterion = nn.CrossEntropyLoss()
+
+    def _reset_group_features(self):
+
+        # see Table 7b)
+        features = self.memory_bank.bank
+        if features is not None:
+            features = features.t().cpu().numpy()
+            kmeans = KMeans(self.n_groups).fit(features)
+            new_features = torch.FloatTensor(kmeans.cluster_centers_).cuda()
+            self.smog.init_groups(new_features)
+
+    def _reset_momentum_weights(self):
+        # see Table 7b)
+        self.backbone_momentum = copy.deepcopy(self.backbone)
+        self.projection_head_momentum = copy.deepcopy(self.projection_head)
+        utils.deactivate_requires_grad(self.backbone_momentum)
+        utils.deactivate_requires_grad(self.projection_head_momentum)
+
+    def training_step(self, batch, batch_idx):
+
+        if batch_idx == 0:
+            # the original paper reset group features and weights every 300
+            # iterations, here we do it once an epoch (TODO)
+            self._reset_group_features()
+            self._reset_momentum_weights()
+            pass
+        else:
+            # update momentum
+            utils.update_momentum(self.backbone, self.backbone_momentum, 0.99)
+            utils.update_momentum(self.projection_head, self.projection_head_momentum, 0.99)
+
+        (x0, x1), _, _ = batch
```

Comment:
I think you forgot to use the `SMoGCollateFunction` :)

Reply:
This is on purpose to keep it comparable with the other strategies. But I'll try it and report back.